### PR TITLE
Services

### PR DESCRIPTION
**Pull request by dhagandev**
Twitter related logic moved to its own service

### CODE REVIEW

File: src/main/java/liuni/resources/LiUniResource.java [hunk 1]
```diff
@@ -28,55 +27,44 @@
...
-        twitterTimeline = null;
-        twitterStatus = null;
+        twitterService = TwitterService.getInstance();
     }
 
     public LiUniResource(LiUniConfig config, int index) {
         this.config = config;
-        twitterTimeline = null;
-        twitterStatus = null;
+        twitterService = TwitterService.getInstance();
```
Comment (kev-chan) [inline src/main/java/liuni/resources/LiUniResource.java]: you are hardcoding the index in the resource class, if you're doing that i'd recommend cleaning up the code

File: src/main/java/liuni/resources/LiUniResource.java [hunk 2]
```diff
@@ -28,55 +27,44 @@
 @Produces(MediaType.APPLICATION_JSON)
 public class LiUniResource {
     private static Logger logger = LoggerFactory.getLogger(LiUniResource.class);
-    private TwitterTimeline twitterTimeline;
-    private TwitterStatus twitterStatus;
     private LiUniConfig config;
+    private TwitterService twitterService;
 
     public LiUniResource() {
```
Comment (kev-chan) [inline src/main/java/liuni/resources/LiUniResource.java]: this is never used, clean it up

File: src/main/java/liuni/resources/LiUniResource.java [hunk 3]
```diff
@@ -1,12 +1,11 @@
 package liuni.resources;
 
 import liuni.LiUniConfig;
-import liuni.TwitterStatus;
-import liuni.TwitterTimeline;
 import liuni.api.ErrorModel;
 import liuni.api.TwitterTimelineModel;
 import com.codahale.metrics.annotation.Timed;
 import liuni.api.TwitterTweetModel;
+import liuni.services.TwitterService;
```
Comment (kev-chan) [inline src/main/java/liuni/resources/LiUniResource.java]: call the <name>Driver class -> <name>Application

File: src/main/java/liuni/services/TwitterService.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import twitter4j.conf.ConfigurationBuilder;
+
+public final class TwitterService {
+    private final static TwitterService INSTANCE = new TwitterService();
+    private static Logger logger = LoggerFactory.getLogger(TwitterService.class);
+
+    private TwitterConfig twitterConfig;
+    private Twitter twitter;
+
+    public static final int TWITTER_CHAR_MAX = 280;
```
Comment (kev-chan) [inline src/main/java/liuni/services/TwitterService.java]: keep all of the static variables together

File: src/main/java/liuni/services/TwitterService.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import twitter4j.ResponseList;
+import twitter4j.Status;
+import twitter4j.Twitter;
+import twitter4j.TwitterException;
+import twitter4j.TwitterFactory;
+import twitter4j.conf.ConfigurationBuilder;
+
+public final class TwitterService {
+    private final static TwitterService INSTANCE = new TwitterService();
+    private static Logger logger = LoggerFactory.getLogger(TwitterService.class);
```
Comment (kev-chan) [inline src/main/java/liuni/services/TwitterService.java]: this can be final as well

File: src/main/java/liuni/services/TwitterService.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            Status status = twitter.updateStatus(text);
+            logger.info("Successfully updated status to [" + status.getText() + "].");
+        }
+        return isOkToPost;
+    }
+
+    // True = No errors; False = Error occurred
+    private boolean textErrorCheck(String text) {
+        char[] convertedArr = text.toCharArray();
+        int statusLength = convertedArr.length;
```
Comment (kev-chan) [inline src/main/java/liuni/services/TwitterService.java]: why is this being converted to a char[]?

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/main/java/liuni/LiUniApplication.java [hunk 7]
```diff
@@ -20,11 +23,10 @@ public static void main(String args[]) {
 
     @Override
     public void run(LiUniConfig config, Environment env) {
-        final LiUniResource resource = new LiUniResource(config, 0);
+        final LiUniResource resource = new LiUniResource(config, index);
```
Comment (kev-chan) [inline src/main/java/liuni/LiUniApplication.java]: why are you passing in the index?

File: src/main/java/liuni/services/TwitterService.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+import twitter4j.ResponseList;
+import twitter4j.Status;
+import twitter4j.Twitter;
+import twitter4j.TwitterException;
+import twitter4j.TwitterFactory;
+import twitter4j.conf.ConfigurationBuilder;
+
+public final class TwitterService {
+    public final static int TWITTER_CHAR_MAX = 280;
+    private final static TwitterService INSTANCE = new TwitterService();
```
Comment (kev-chan) [inline src/main/java/liuni/services/TwitterService.java]: lazy load this

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/main/java/liuni/LiUniApplication.java (hunk 7 above)
Comment (dhagandev) [reply src/main/java/liuni/LiUniApplication.java]: Allows you to switch between users in the yml file, is there a better way to be indicating which user you should use?

Comment (kev-chan) [reply src/main/java/liuni/LiUniApplication.java]: there is, give it some thought

File: src/main/java/liuni/services/TwitterService.java [hunk 9]
```diff
@@ -12,8 +12,8 @@
 
 public final class TwitterService {
     public final static int TWITTER_CHAR_MAX = 280;
-    private final static TwitterService INSTANCE = new TwitterService();
     private final static Logger logger = LoggerFactory.getLogger(TwitterService.class);
+    private static volatile TwitterService INSTANCE = null;
```
Comment (kev-chan) [inline src/main/java/liuni/services/TwitterService.java]: you don't need volatile/synchronized

File: application-example.yml [hunk 10]
```diff
@@ -1,9 +1,11 @@
 twitter:
-  - userName:
-    consumerKey:
-    consumerSecret:
-    accessToken:
-    accessSecret:
+  defaultuser: 0
+  users:
+    - userName:
```
Comment (kev-chan) [inline application-example.yml]: come up with a better variable name for what this value actual represents, variable name is confusing

File: application-example.yml [hunk 11]
```diff
@@ -1,9 +1,11 @@
 twitter:
-  - userName:
-    consumerKey:
-    consumerSecret:
-    accessToken:
-    accessSecret:
+  defaultuser: 0
```
Comment (kev-chan) [inline application-example.yml]: same for this one, better variable name

**Review by kev-chan — CHANGES_REQUESTED** [review]
update the error message when auth credentials to be more professional.  no Chromedino

File: application-example.yml [hunk 12]
```diff
@@ -1,9 +1,10 @@
 twitter:
-  - userName:
-    consumerKey:
-    consumerSecret:
-    accessToken:
-    accessSecret:
+  defaultUserIndex: 0
+  users:
```
Comment (kev-chan) [inline application-example.yml]: call this account instead of user

**Review by kev-chan — CHANGES_REQUESTED** [review]

**Pull request merged**